tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc

```diff
@@ -671,14 +672,19 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
     return;
   }
 
-  if (!use_implicit_batch_ && has_dynamic_shape_input_) {
+  if (!use_implicit_batch_ && has_dynamic_shape_input_ ||
+      cache_res->profiles_.HasShapeTensor()) {
+    // Append shape values as TensorShapes
+    OP_REQUIRES_OK_ASYNC(ctx, cache_res->profiles_.CollectShapeValues(ctx),
+                         *helper);
     if (profile_generation_mode_) {
       // Collecting new shapes for profiles can be only done once. After the
       // shapes are converted to TRT profiles, no shapes can be collected
       // anymore.
-      OP_REQUIRES(ctx, cache_res->profiles_.GetNumProfiles() == 0,
-                  errors::Unimplemented("Cannot collect new shapes when "
-                                        "profiles are already created."));
+      OP_REQUIRES_ASYNC(ctx, cache_res->profiles_.GetNumProfiles() == 0,
+                        errors::Unimplemented("Cannot collect new shapes when "
+                                              "profiles are already created."),
+                        *helper);
       // Just collect the input shape info and return. The shapes are used to
       // generate optimization profiles during engine creation.
       cache_res->profiles_.AddShape(input_concrete_shapes);
```

Comment:
CollectShapeValues preprares actual_shape_values for AddShape. From this code block, it is possible that AddShape won't be called after CollectShapeValues. It is not straightforward to understand.
Can we combine CollectShapeValues and AddShape and remove actual_shape_values?